File: iOS-ObjC/watf-bank/Http.m
#import "Http.h"
#import "Util.h"

@interface Http() <NSURLSessionDelegate>
@end

@implementation Http

static NSString *ip;

-(void)URLSession:(NSURLSession *)session didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition, NSURLCredential *))completionHandler
{
    completionHandler(NSURLSessionAuthChallengeUseCredential, [NSURLCredential credentialForTrust:challenge.protectionSpace.serverTrust]);
}

+(void)setIp:(NSString *)ipp
{
    if(ip==nil)
    {
        ip = [[NSString alloc] init];
    }
    ip = ipp;
}

+(NSString*)getIp
{
    if(ip==nil)
    {
        [self setIp:@"192.168.1.1"];
    }
    return ip;
}

+(void)post:(NSString*)api :(NSDictionary*)data :(void (^)(NSDictionary *res))completion
{
    NSString *url = [NSString stringWithFormat:@"https://%@:5000/",ip];
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"%@%@",url,api]]];
    NSError *e = nil;
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/json" forHTTPHeaderField:@"content-type"];
    [request setHTTPBody:[NSJSONSerialization dataWithJSONObject:data options:NSJSONWritingPrettyPrinted error:&e]];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration] delegate:[self alloc] delegateQueue:[NSOperationQueue mainQueue]];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data,
                                                                           NSURLResponse * _Nullable response,
                                                                           NSError * _Nullable error) {
        NSError *e = nil;
        if(data==nil)
        {
            completion(nil);
            return;
        }
        NSDictionary *res = [NSJSONSerialization JSONObjectWithData:data options:0 error:&e];
        completion(res);
    }];
    [task resume];
}

@end

File: iOS-Swift/watf-bank/Http.swift
import Foundation
import UIKit

class Http: NSObject, URLSessionDelegate {
    
    private static var ip = "127.0.0.1:5000"
    
    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Swift.Void) {
        completionHandler(URLSession.AuthChallengeDisposition.useCredential, URLCredential(trust: challenge.protectionSpace.serverTrust!))
    }
    
    func post(_ payload: [String: String],_ api: String, completionHandler:@escaping ([String: Any]) -> ()) {
        var request = URLRequest(url: URL(string: "https://\(Http.getIp())/\(api)")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload, options: .prettyPrinted )
        let config = URLSessionConfiguration.default
        let session = URLSession(configuration: config, delegate: self, delegateQueue: OperationQueue.main)
        let task = session.dataTask(with: request) { data, response, error in
            guard let data = data, error == nil else {
                completionHandler(["error":"Connection Failed"])
                return
            }
            let r = response as! HTTPURLResponse
            if r.statusCode != 200 {
                completionHandler(["error":"\(r.statusCode):\(HTTPURLResponse.localizedString(forStatusCode: r.statusCode))"])
                return
            }
            let tmp = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: String]
            if tmp! == nil {
                let res = try? JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
                completionHandler(res!)
            }
            else {
                let res = try? JSONSerialization.jsonObject(with: data, options: []) as! [String: String]
                completionHandler(res!)
            }
        }
        task.resume()
    }
    
    static func setIp(_ ip: String) {
        self.ip = ip
    }
    
    static func getIp() -> String {
        return KeyChain.load("ip") ?? ip
    }
}
